services/inc/scope_guard.h

```diff
@@ -15,16 +15,124 @@
  * License along with this library; if not, see <http://www.gnu.org/licenses/>.
  */
 
+/**
+ * @file scope_guard.h
+ *
+ * @section scope_guards Scope guards
+ *
+ * This header file defines a set of macros that simplify RAII-style resource management in the
+ * Device OS code.
+ *
+ * Consider the following example:
+ * ```
+ * int parseStream(Stream* stream) {
+ *     // Allocate a buffer
+ *     const size_t bufferSize = 1024;
+ *     char* buffer = (char*)malloc(bufferSize);
+ *     if (!buffer) {
+ *         return SYSTEM_ERROR_NO_MEMORY;
+ *     }
+ *     // Read the first chunk of the data
+ *     int result = stream->read(buffer, bufferSize);
+ *     if (result < 0) {
+ *         free(buffer);
+ *         return result;
+ *     }
+ *     ...
+ *     // Read the second chunk of the data
+ *     result = stream->read(buffer, bufferSize);
+ *     if (result < 0) {
+ *         free(buffer);
+ *         return result;
+ *     }
+ *     ...
+ *     // Free the buffer
+ *     free(buffer);
+ *     return 0;
+ * }
+ * ```
+ * Using the `SCOPE_GUARD()` macro together with the @ref check_macros "check macros", the above
+ * code can be rewritten in a less error-prone way:
+ * ```
+ * int parseStream(Stream* stream) {
+ *     // Allocate a buffer
+ *     const size_t bufferSize = 1024;
+ *     char* buffer = (char*)malloc(bufferSize);
+ *     CHECK_TRUE(buffer, SYSTEM_ERROR_NO_MEMORY);
+ *     // Declare a scope guard
+ *     SCOPE_GUARD({
+ *         free(buffer);
+ *     });
+ *     // Read the first chunk of the data
+ *     CHECK(stream->read(buffer, bufferSize));
+ *     ...
+ *     // Read the second chunk of the data
+ *     CHECK(stream->read(buffer, bufferSize));
+ *     ...
+ *     return 0;
+ * }
+ * ```
+ * Internally, `SCOPE_GUARD()` declares an object that schedules an arbitrary user-provided code
+ * for execution at the end of the enclosing scope of the object. In the above example, the buffer
+ * allocated via `malloc()` will be automatically freed by the scope guard when the function returns
+ * either successfully or with an error.
+ *
```

Comment:
If there are any ways that this macro syntax could be abused or used incorrectly, it should be noted here.